### PR TITLE
[CARBONDATA-2798] Fix Dictionary_Include for ComplexDataType

### PR DESCRIPTION
**Pull request by Indhumathi27**
**Problem1:**
Select Filter is throwing BufferUnderFlow Exception as cardinality is filled for Non-Dictionary columns.
**Solution:**
Check if a complex column has Encoding => Dictionary and fill cardinality for that column only.

**Problem2:**
Transactional Table is throwing NullPointerException if csv fileheader is not proper.
**Solution:**
Throw CarbonDataLoadingException if csv fileheader is not proper.
 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [ ] Testing done
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7607/

File: processing/src/main/java/org/apache/carbondata/processing/datatypes/ArrayDataType.java [hunk 1]
```diff
@@ -172,7 +191,7 @@ public void setSurrogateIndex(int surrIndex) {
 
   @Override
   public void fillCardinality(List<Integer> dimCardWithComplex) {
-    if (children.getIsColumnDictionary()) {
+    if (this.isDictionaryColumn) {
```
Comment (kunal642) [inline processing/src/main/java/org/apache/carbondata/processing/datatypes/ArrayDataType.java]: return isDictionaryColumn from getIsColumnDictionary() also.

**Comment by kunal642** [comment]
Please add a test case to verify the fix

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7615/

**Comment by brijoobopanna** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7628/

**Comment by Indhumathi27** [comment]
Retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7638/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6358/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6378/

File: processing/src/main/java/org/apache/carbondata/processing/loading/model/LoadOption.java [hunk 2]
```diff
@@ -236,7 +236,7 @@
       }
     }
 
-    if (carbonLoadModel.isCarbonTransactionalTable() && !CarbonDataProcessorUtil
+    if (hadoopConf != null && !CarbonDataProcessorUtil
```
Comment (kunal642) [inline processing/src/main/java/org/apache/carbondata/processing/loading/model/LoadOption.java]: why hadoopConf null check is needed?

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6071/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7658/

**Comment by brijoobopanna** [comment]
retest this please

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6392/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7672/

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/createTable/TestNonTransactionalCarbonTable.scala [hunk 3]
```diff
@@ -52,10 +52,11 @@ import org.apache.carbondata.core.datastore.page.encoding.DefaultEncodingFactory
 import org.apache.carbondata.core.metadata.ColumnarFormatVersion
 import org.apache.carbondata.core.metadata.datatype.DataTypes
 import org.apache.carbondata.core.util.{CarbonProperties, CarbonUtil, DataFileFooterConverterV3}
+import org.apache.carbondata.processing.loading.exception.CarbonDataLoadingException
 import org.apache.carbondata.sdk.file._
 
 
-class TestNonTransactionalCarbonTable extends QueryTest with BeforeAndAfterAll {
+class TestNonTransactiogitnalCarbonTable extends QueryTest with BeforeAndAfterAll {
```
Comment (kunal642) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/spark/testsuite/createTable/TestNonTransactionalCarbonTable.scala]: Please revert this change

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6402/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/7684/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/6408/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/6085/

**Comment by kunal642** [comment]
LGTM
